analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CollectionTracker.cs

```diff
@@ -146,4 +146,67 @@ public static ProgramState LearnFrom(ProgramState state, IPropertyReferenceOpera
         }
         return state;
     }
+
+    public static ProgramState LearnFrom(ProgramState state, IInvocationOperationWrapper invocation)
+    {
+        if (EnumerableCountConstraint(state, invocation) is { } constraint)
+        {
+            return state.SetOperationConstraint(invocation, constraint);
+        }
+        var targetMethod = invocation.TargetMethod;
+        if (invocation.Instance is { } instance && instance.Type.DerivesOrImplementsAny(CollectionTypes))
+        {
+            var symbolValue = state[instance] ?? SymbolicValue.Empty;
+            if (AddMethods.Contains(targetMethod.Name))
+            {
+                return SetOperationAndSymbolValue(symbolValue.WithConstraint(CollectionConstraint.NotEmpty));
+            }
+            else if (RemoveMethods.Contains(targetMethod.Name))
+            {
+                return SetOperationAndSymbolValue(symbolValue.WithoutConstraint(CollectionConstraint.NotEmpty));
+            }
+            else if (targetMethod.Name == nameof(ICollection<int>.Clear))
+            {
+                return SetOperationAndSymbolValue(symbolValue.WithConstraint(CollectionConstraint.Empty));
+            }
+        }
+        return state;
+
+        ProgramState SetOperationAndSymbolValue(SymbolicValue value)
+        {
+            if (instance.TrackedSymbol(state) is { } symbol)
+            {
+                state = state.SetSymbolValue(symbol, value);
+            }
+            return state.SetOperationValue(instance, value);
+        }
+    }
+
+    private static NumberConstraint EnumerableCountConstraint(ProgramState state, IInvocationOperationWrapper invocation)
+    {
+        if (invocation.TargetMethod.Is(KnownType.System_Linq_Enumerable, nameof(Enumerable.Count)))
+        {
+            var instance = invocation.Instance ?? invocation.Arguments[0].ToArgument().Value;
+            if (instance.TrackedSymbol(state) is { } symbol
+                && state[symbol]?.Constraint<CollectionConstraint>() is { } collection)
+            {
+                if (collection == CollectionConstraint.Empty)
+                {
+                    return NumberConstraint.From(0);
+                }
+                else if (!HasFilteringPredicate()) // nonEmpty.Count(predicate) can be Empty or NotEmpty
+                {
+                    return NumberConstraint.From(1, null);
+                }
```

Comment:
This logic (maybe even more of this method) should be shared with `LearnFrom(ProgramState, IPropertyReferenceOperationWrapper, ISymbol)`.

Reply:
ok fine. The shared logic is very short anyway and it's only two methods using it.